dgram: setMulticastInterface() outgoing interface selection

lostnet opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [X] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

dgram
##### Description of change

<!-- Provide a description of the change below this comment. -->

Add wrapper for uv's uv_udp_set_multicast_interface which provides the
sender side mechanism to explicitly select an interface. (The
equivalent receiver side mechanism is the optional 2nd argument of
addMembership().)

santigimeno commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,31 @@
+    socket2.setMulticastInterface('::');
+  } catch (e) {
+    thrown = true;
+  }
```
wrapping this with `assert.throws` will allow to remove the `thrown` variable

santigimeno commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,31 @@
+
+  //close the socket
+  socket.close();
+});
```
Add `common.mustCall()` to the `listening` callback?

santigimeno commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,31 @@
+
+socket2.bind(0);
+socket2.on('listening', function() {
+
```
Same here.

lostnet replied:
For the above 3, I was using: test-dgram-multicast-setTTL.js as a template for requirements and style as it is the nearest test. I will make these updates using test-http-response-statuscode.js as a template.

santigimeno commented on src/udp_wrap.cc:
```diff
@@ -213,6 +214,25 @@ X(SetMulticastLoopback, uv_udp_set_multicast_loop)
+  }
+
+  int err = uv_udp_set_multicast_interface(&wrap->handle_,
+                                  iface_cstr);
```
I think the function arguments should be aligned

lostnet replied:
style alignment line 233? I don't see any pointer alignment issues.

santigimeno replied:
> style alignment line 233?

Yes, I meant that :) :
```js
int err = uv_udp_set_multicast_interface(&wrap->handle_,
                                         iface_cstr);
```

santigimeno commented:
I don't know how difficult it would be to test, but would it be possible to test that the sender is indeed sending the multicast messages using the chosen interface?

lostnet commented:
> I don't know how difficult it would be to test, but would it be possible to test that the sender is indeed sending the multicast messages using the chosen interface?

It is very easy to test on a system where you can copy output from an `ifconfig -a` into your script and at least 2 real interfaces have been setup. For automated tests, its rather hard to get anyone to do the prerequisite manual system configuration.

lostnet commented:
Thanks santigimeno, I've made improvements corresponding to each of your comments and verified "make -j4 test" against the new version. Please let me know how it looks.

cjihrig commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,26 @@
+socket2.bind(0);
+socket2.on('listening', common.mustCall(() => {
+
+  //Try to set with an invalid Interface address (wrong address class)
```
Space after `//`. Also, Interface doesn't need to be capitalized.

cjihrig commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,26 @@
+const assert = require('assert');
+const dgram = require('dgram');
+const socket = dgram.createSocket('udp4');
+const socket2 = dgram.createSocket('udp4');
```
Instead of having `socket`, `socket2`, ..., `socketN`, we've been moving toward using block scoping to isolate tests. So:

```javascript
{
  const socket = dgram.createSocket('udp4');
  ...
}

{
  const socket = dgram.createSocket('udp4');
  ...
}
```

cjihrig commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,23 @@ added: v0.6.9
 
+### socket.setMulticastInterface(multicastInterface)
+<!-- YAML
+added: v7.0.0
```
Instead of `v7.0.0`, please use `REPLACEME`. That value will be replaced during the release.

cjihrig commented on doc/api/dgram.md:
````diff
@@ -299,6 +299,23 @@ added: v0.6.9
+```js
+const s = dgram.createSocket('udp6');
+s.bind(1234, () => {
+        s.setMulticastInterface('::%eth1');
````
This indentation is off.

yorkie commented on doc/api/dgram.md:
````diff
@@ -299,6 +299,23 @@ added: v0.6.9
+Sets the outgoing multicast interface using the provided IP. For IPv6, the addresses should use explicit scope to indicate the interface, as in the following example:
+ 
+```js
+const s = dgram.createSocket('udp6');
````
Use the more complete name `socket` instead of `s`?

cjihrig commented on lib/dgram.js:
```diff
@@ -495,6 +495,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+    throw new Error('interface address must be specified');
+  }
+
+  var err = this._handle.setMulticastInterface(interfaceAddress);
```
Use `const` please.

cjihrig commented on src/udp_wrap.cc:
```diff
@@ -213,6 +214,25 @@ X(SetMulticastLoopback, uv_udp_set_multicast_loop)
+
+  const char* iface_cstr = *iface;
+  if (args[0]->IsUndefined() || args[0]->IsNull()) {
+      iface_cstr = nullptr;
```
Two space indentation please.

cjihrig commented:
Perhaps you could add a test similar to `test/internet/test-dgram-multicast-multi-process.js`.

lostnet commented:
Thank you @cjihrig and @yorkie, I agree with, and believe I have addressed all the comments in this second set of review changes.

On test-dgram-multicast-multi-process.js:
I've added default interface selection and explicitly to the membership in that test to show that they do no harm. But the only ways I've seen to test that this kind of network setting isn't always doing a NO-OP is with test frameworks that have  either a privileged system configuration stage or a configuration verification stage where they decide if they can test opportunistically. (The test framework has to know what a real non-default interface's IP is and that it isn't physically on the same physical network as the default network or all packets get received simply since there isn't an alternative.)

Please let me know what you think, and if I've missed or misunderstood anything on the other changes,
Will

yorkie commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,32 @@
+'use strict';
```
filename in test/parallel should be in lower-case?

lostnet replied:
I like the order with regards to dgram-multicast for tab completion and relatedness, so maybe:
test/parallel/test-dgram-multicast-set-interface.js ?
(set_if is the only lowercase I would parse as an _IF ioctl and IMO mixing '-'&'_' is worse than mixed case.)

yorkie commented on src/udp_wrap.cc:
```diff
@@ -213,6 +214,25 @@ X(SetMulticastLoopback, uv_udp_set_multicast_loop)
+  node::Utf8Value iface(args.GetIsolate(), args[0]);
+
+  const char* iface_cstr = *iface;
+  if (args[0]->IsUndefined() || args[0]->IsNull()) {
```
`args[0]->IsUndefined()` seems to be untouchable because we checked the address at the [JS level](https://github.com/nodejs/node/pull/7855/files#diff-5c0ca4f44209cc4cc68c4dccadbd4a07R501).

jasnell replied:
Agreed... since the js code requires that the `interfaceAddress` be specified, or, if we switch to requiring it to be a string, I would recommend simply switching these to `CHECK` assertions... e.g. 

```js
CHECK(!args[0]->IsUndefined());
CHECK(!args[0]->IsNull());
Utf8Value iface(args.GetIsolate(), args[0]);
int err = uv_udp_set_multicast_interface(&wrap->handle_, *iface);
```

Also note: it should be unnecessary to use the `node::` prefix on the Utf8Value decl

lostnet replied:
> args[0]->IsUndefined() seems to be untouchable because we checked the address at the JS level.

I think the JS checks are just a courtesy to let programs that don't do anything seriously wrong deal with an exception and live, I can bypass it and start triggering these paths, i,e:

      socket.on('listening', () => {
          socket._handle.setMulticastInterface();

so to verify there's no unaccountable corruption, I think this side needs to force an abort whenever it hasn't verified that the caller has been railroaded into a correct path.

How about:

       CHECK_EQ(args.Length(), 1);
       CHECK(args[0]->IsString());
       Utf8Value iface(args.GetIsolate(), args[0]);
       int err = uv_udp_set_multicast_interface(&wrap->handle_, *iface);

cjihrig commented on doc/api/dgram.md:
````diff
@@ -299,6 +299,23 @@ added: v0.6.9
+ 
+```js
+const socket = dgram.createSocket('udp6');
+socket.bind(1234, () => {
````
Could you add a blank line here.

cjihrig commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,23 @@ added: v0.6.9
+socket.bind(1234, () => {
+  socket.setMulticastInterface('::%eth1');
+});
+
```
And remove this blank line.

cjihrig commented on lib/dgram.js:
```diff
@@ -495,6 +495,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+Socket.prototype.setMulticastInterface = function(interfaceAddress) {
+  this._healthCheck();
+
+  if (!interfaceAddress) {
```
Instead of checking for falsey values, maybe this should check for strings explicitly.

jasnell replied:
either that or `interfaceAddress` should be explicitly coerced, e.g. `interfaceAddress = String(interfaceAddress)`

lostnet replied:
I am going to go with the typeof check since String(undefined) would be relying on "undefined" being a (hopefully) unparseable address and obfuscate which type of error was made. (It will also align with the change I've made in the wrapper based on another comment eliminating any window for callers of this function to trigger an abort instead of an exception.)

cjihrig commented on lib/dgram.js:
```diff
@@ -495,6 +495,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+  this._healthCheck();
+
+  if (!interfaceAddress) {
+    throw new Error('interface address must be specified');
```
That would make this a `TypeError`.

jasnell replied:
Also, s/interface address/interfaceAddress

cjihrig commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,32 @@
+  socket.on('listening', common.mustCall(() => {
+    socket.setMulticastInterface('0.0.0.0');
+
+    // close the socket
```
This comment seems pretty unnecessary :-)

cjihrig commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,32 @@
+      socket.setMulticastInterface('::');
+    }));
+
+    // close the socket
```
Same here.

jasnell commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,23 @@ added: v0.6.9
+* `multicastInterface` {String}
+
+Sets the outgoing multicast interface using the provided IP. For IPv6, the addresses should use explicit scope to indicate the interface, as in the following example:
+ 
```
This should likely include some additional detail about *when* this method should be called and what the side effects are.

lostnet replied:
I've added a significant amount of detail (and hopefully used subsections correctly?) But I've intentionally left the question of exactly when a socket is first ready unanswered since it would be redundant to repeat libuv's unusual need to bind before socket operations for every operation.

jasnell commented on test/parallel/test-dgram-multicast-setIf.js:
```diff
@@ -0,0 +1,32 @@
+    // close the socket
+    socket.close();
+  }));
+}
```
This test case should also test the error conditions when invalid interface names are given, the argument is not provided at all, or the method is called at the wrong time, etc.

lostnet replied:
I've added tests along these lines. It should be tested by someone with windows and perhaps OsX as which error or whether something is an error is often left open to interpretation in the specs.

cjihrig replied:
Our CI will test these platforms.

jasnell commented:
Left a few comments. Thank you for doing this! It's almost there!

lostnet commented:
Thank you @yorkie, @cjihrig and @jasnell for you feedback on revision 2, I believe I have addressed them all, but I would appreciate any feedback, especially on the additions to docs and tests.

cjihrig commented on lib/dgram.js:
```diff
@@ -495,6 +495,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+  this._healthCheck();
+
+  if (typeof interfaceAddress !== 'string') {
+    throw new TypeError('interfaceAddress must be a String');
```
String -> string

cjihrig commented on src/udp_wrap.cc:
```diff
@@ -213,6 +214,23 @@ X(SetMulticastLoopback, uv_udp_set_multicast_loop)
+  const char* iface_cstr = *iface;
+
+  int err = uv_udp_set_multicast_interface(&wrap->handle_,
+                                           iface_cstr);
```
If this can fit on the previous line, can you please move it.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+
+  socket.bind(0);
+  socket.on('listening', common.mustCall(() => {
+
```
Please remove this blank line.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+
+  socket.bind(0);
+  socket.on('listening', common.mustCall(() => {
+
```
Same with this blank line, and others.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+  socket.on('listening', common.mustCall(() => {
+
+    // Try to set with an invalid interfaceAddress (wrong address class)
+    assert.throws(common.mustCall(() => {
```
You don't need `common.mustCall()` around functions passed to `assert.throws()`. If the function isn't called, then it won't throw, and the assertion will fail. `common.mustCall()` is primarily used to ensure that asynchronous callback functions are actually called.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+    // Try to set with an invalid interfaceAddress (wrong address class)
+    assert.throws(common.mustCall(() => {
+      socket.setMulticastInterface('::');
+    }));
```
Is there a second argument you can pass to `assert.throws()` that will properly identify the expected error?

lostnet replied:
In this case, libuv currently makes the call with the other address family so each OS could pick a slightly different error from among any unsupported/unavailable/general ones, (and libuv could also later refactor this into a EINVAL or other error instead of making the call.)

I used this test since I think it is worth the risk to determine that no OS with a little too much compatibility code tries to warp this into a success, but this is one of the errors where the exact response could vary quite a bit (which I added to the docs).

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+    assert.throws(common.mustCall(() => {
+      socket.setMulticastInterface(1);
+    }),
+    /TypeError/);
```
Please move this to the previous line.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+    // Try to set with an invalid interfaceAddress (non-unicast)
+    assert.throws(common.mustCall(() => {
+      socket.setMulticastInterface('224.0.0.2');
+    }));
```
Again, a second argument that identifies the error would be preferred here.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+    assert.throws(common.mustCall(() => {
+      socket.setMulticastInterface(String(undefined));
+    }),
+    /EINVAL/);
```
Move this to the previous line please.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -0,0 +1,122 @@
+    assert.throws(common.mustCall(() => {
+      socket.setMulticastInterface('');
+    }),
+    /EINVAL/);
```
Please move to the previous line.

cjihrig commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+
+* `multicastInterface` {String}
+
+Sets the default outgoing multicast interface of the socket to a chosen interface or back to system interface selection. The `multicastInterface` must be a valid string representation of an IP from the socket's family.
```
Please wrap all lines at 80 characters.

cjihrig commented on lib/dgram.js:
```diff
@@ -495,6 +495,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+Socket.prototype.setMulticastInterface = function(interfaceAddress) {
+  this._healthCheck();
+
+  if (typeof interfaceAddress !== 'string') {
```
Would we be able to get away with checking for IP addresses only here (`net.isIP()`)?

lostnet replied:
net.isIP() makes it look like no checks are necessary either in JS or the wrapper:

Its existing tests don't cover Scope, so I added some:
assert.equal(net.isIP('::%eth1'), 6);
assert.equal(net.isIP('::1%1'), 6);
These passed leading me to verify it is using a C implementation to get to libuv:
https://github.com/nodejs/node/blob/f8d3f6fbf4b0471677a0e4becf458a186fb2e4de/src/cares_wrap.cc#L34

yet, this doesn't need any wrapper:
https://github.com/nodejs/node/blob/863952ebadd4909a66bb9da7868bf75bbbe1462d/lib/net.js#L1581
and still passes:
assert.equal(net.isIP(), 0);

If net.isIP() is correct then no check on interfaceAddress in JS and removing the argument checks should be fine?

yorkie replied:
@lostnet would you please cover the missing parts for `net.isIP()` as you described in another patch?

jasnell commented:
LGTM with the various nits @cjihrig pointed out addressed. Thank you!

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+
+For IPv4 sockets, this should be the IP configured for the desired physical interface. All packets sent to multicast on the socket will be sent on the interface determined by the most recent successful use of this call.
+
+For IPv6 sockets, `multicastInterface` should include a Scope ("IP%Scope") to indicate the interface as in the examples that follow. In IPv6, individual send calls can also use explicit Scope in addresses, so only packets sent to a multicast address without specifying an explicit Scope are affected by the most recent successful use of this call.
```
1. `s/Scope/scope/g`?
2. "individual send calls" -> "individual `send` calls" might be better to understand which specify the `send` is a operation/function, not the verb.

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+For IPv6 sockets, `multicastInterface` should include a Scope ("IP%Scope") to indicate the interface as in the examples that follow. In IPv6, individual send calls can also use explicit Scope in addresses, so only packets sent to a multicast address without specifying an explicit Scope are affected by the most recent successful use of this call.
+ 
+#### Examples: IPv6 Outgoing Multicast Interface
+On most systems, where Scope format uses the interface name:
```
That would be nice if we could document what is the "scope" and "scope format" :-)

lostnet replied:
It seems odd to me that there isn't a very good language neutral description of them to reference since every environment that runs on windows and one *nix needs the same documentation. Looking at libuv, the topic is skipped all together.

But I think it would make more sense to do a follow up patch to add scope_name to [os.networkInterfaces()](https://nodejs.org/api/os.html#os_os_networkinterfaces) and just link to whatever docs can be handled there.

yorkie replied:
Hmm, this documents is writing for Node.js user who might not be too familiar with this even at operating system and the libuv.

Basically I think here we should notice reader that:

```md
Note: all `Scope` in this section represents the [IPv6 address scope](https://en.wikipedia.org/wiki/IPv6_address#IPv6_address_scopes)
```

Then any confuser will read the wiki if he wants to know :-)

lostnet replied:
It is actually referred to as a zone index in that page:
https://en.wikipedia.org/wiki/IPv6_address#Link-local_addresses_and_zone_indices

but in programming materials it is more confusingly referred to as the scope identifier, i.e.:
http://tldp.org/HOWTO/Linux+IPv6-HOWTO/chapter-section-using-api.html#AEN3142 (section 23.1.1.2)

yorkie commented on doc/api/dgram.md:
````diff
@@ -299,6 +299,62 @@ added: v0.6.9
+});
+```
+
+On windows, where Scope format uses an interface number:
````
`s/windows/Windows/g`?

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+
+#### Call Results
+
+A call on a socket that is not ready to send or no longer open may cause a *Not running* Error Exception.
```
The `Error` seems to be duplicated with `Exception` in semantics, so propose choosing one of them, and +1 with `Error` because it is a real type in v8.

lostnet replied:
I've dressed up `Error` classes and removed uses of `Exception` but made sure 'thrown' is always used as the verb. I want to make sure it is clear I am talking about an Error to be caught in an Exception since I think a reader could be expecting them via events (i.e. as `socket.bind` does) or the return value (for those thinking in terms of standard socket operations.)

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+
+A call on a socket that is not ready to send or no longer open may cause a *Not running* Error Exception.
+
+If `multicastInterface` can not be parsed into an IP then an *EINVAL* Error Exception is thrown.
```
Ditto, `s/Error Exception/System Error/` to keep consistent with our [docs#errors.html](https://github.com/nodejs/node/blob/master/doc/api/errors.md#errors).

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,62 @@ added: v0.6.9
+
+If `multicastInterface` can not be parsed into an IP then an *EINVAL* Error Exception is thrown.
+
+On IPv4, if `multicastInterface` is a valid Address but does not match any interface, or if the address does not match the family then system provided errors such as *EADDRNOTAVAIL* or *EPROTONOSUP* are indicated in the Error Exception.
```
`s/Address/address/g`?

lostnet commented:
Thank you @cjihrig @yorkie and @jasnell for your reviews and comments of update #3, I believe I have now integrated all of them (and opened a separate one for net.isIP).

lo test:
I have also done some testing on loopback and determined that I can add a test like the one requested that will at least work on Linux in v4 using loopback as the 2nd interface. The CI tests can determine whether it functions on other OSes, and I have left my v6 support in if anyone would like to investigate v6 on other OSes,  but I think loopback can not work for this test on Linux since loopback is not a separable scope from scope 0.

yorkie commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,84 @@ added: v0.6.9
+
+* `multicastInterface` {String}
+
+*Note- All references to scope in this section are refering to
```
`Note:`?

jasnell replied:
This is fine for now, I think.. I'll be going through all the docs here soon to standardize the formats of the various `Note:` instances

yorkie commented:
The documentation changes looks good to me except for a nit @lostnet Thank you :-)

jasnell commented:
@nodejs/ctc... would appreciate a few more eyes on this! thank you! :-)

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3602/

jasnell commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+    reuseAddr: true
+  });
+
+  // don't bind the address explicitly for sending
```
Extremely minor nit: can you upper case the first letter in the comments and add punctuation?

jasnell commented:
LGTM if CI is green.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+const Buffer = require('buffer').Buffer;
+const fork = require('child_process').fork;
+const MULTICASTS = {
+  IPv4: [ '224.0.0.115', '224.0.0.116'],
```
Can you drop the space after `[`

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+const fork = require('child_process').fork;
+const MULTICASTS = {
+  IPv4: [ '224.0.0.115', '224.0.0.116'],
+  IPv6: [ 'ff02::1:115', 'ff02::1:116']
```
Same thing here.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+const UDP = {IPv4: 'udp4', IPv6: 'udp6'};
+
+const TIMEOUT = common.platformTimeout(5000);
+const NOW = +new Date();
```
Can you use `Date.now()` here?

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+   newAddr: LOOPBACK[FAM]}
+];
+
+if (common.inFreeBSDJail) {
```
Can you move this up in the code.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+// Take the first non-internal interface as the address for interfaceing.
+// Ideally, this should check for whether or not an interface is set up for
+// BROADCAST and favor internal/private interfaces.
+get_interfaceAddress: for (var name in networkInterfaces) {
```
Could you substitute `var` with `let`/`const` in the tests.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+    process.exit(1);
+  }, TIMEOUT);
+
+  //launch child processes
```
Space after `//` and capitalize launch.

cjihrig replied:
Same formatting on the other comments too please.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+  }, TIMEOUT);
+
+  //launch child processes
+  for (var x = 0; x < listeners; x++) {
```
Please use `i`, `j`, etc. for iterator variables.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+    const msg = messages[i++];
+
+    if (!msg) {
+      try { sendSocket.close(); } catch (e) {}
```
What is the purpose of the `try...catch` here? Seems like we should have better control over errors in our tests.

lostnet replied:
I'll remove this since I think it is worth seeing anything from the sender given the nature of this test. I think the reason it was there is since the parent is both the sender and responsible for cleanup, so it is a little more likely a failure in send side will bleed into subsequent tests by leaving orphans.

cjihrig commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,275 @@
+    //HACK: Wait to exit the process to ensure that the parent
+    //process has had time to receive all messages via process.send()
+    //This may be indicitave of some other issue.
+    setTimeout(function() {
```
This might be a bad idea. Hacks like this almost always result in flaky tests in the future. We need a more reliable way to do this.

lostnet replied:
I've refactored this to close the socket before sending the message and exit in the callback from the message. But unfortunately ths kind of a timing issue doesn't look likely yo be responsible for any of the CI failures I've started to investigate.

lostnet commented:
Thank you @jasnell and @yorkie for your feedback on the changes from the 4th review and for kicking off the CI tests. I have integrated the changes and added some fixes for the CI failures:

- For Windows, I have relaxed the address family mismatch test to be purely informational in terms of asserts. I will need to investigate whether there is a problem to follow up with on with libuv, but I don't think the wrapper should play any role.

- For OsX, I have chosen a scope/zone name of lo0 so OsX has a scope name it recognizes as either mapping to 0 and/or binding to scope 0 seems to be done with a strange interpretation on OsX. Other Oses may not have a loopback named 'lo0' (or in window's case will use numbers) but they should pass on the basis of the original test.

- For Linux et al, I have added an additional 2 * {messages, workers, multicast addresses} and started the test with a pretest of multicast. Results of CI tests with these changes should give me two groupings to investigate one of which is purely related to multicast or how I have set it up and the other to the actual setMulticastInterface/loopback behavior I am trying to test.

Many Thanks!
Will

targos commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3661/

lostnet commented:
Thanks @targos for running the tests. I have investigated smartos, raspbian wheezy and windows and made some workarounds:

  - Raspbian - lacked IPv6 support by default, I now dropout before IPv6 tests if a NO-OP test throws an exception.

  - Windows has some kind of [issue binding to multicast addresses](http://erlang.org/pipermail/erlang-questions/2014-May/079032.html), I've changed it to bind to any and separate the workers by port.

  - smartos looks like it still has one of the loopback virtualization bug that was probably present in Solaris at the time of the Indiana fork. I've blocked it from running the loopback test and added a comment with the manual test it needs to pass.

Not addressed:

- For fedora/redhat, I am not sure what is going on since fedora failures jumped between versions 22 and 23. If there are multiple machines of each version (or the tests are run very close to  boot up?) then it may be explained by a default firewall that is blocking multicast. I suppose another possibility would be that they have a source address selection problem.

- AIX: I don't understand the output, I think the tests were canceled?

yorkie commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -73,6 +73,20 @@ const dgram = require('dgram');
+    common.skip(`IPv6 sockets are missing or incompatible: ${e}`);
+    return;
+  }
+}
```
The pair of brackets is unnecessary because you put all lines into `try-catch`, right?

cjihrig commented on test/parallel/test-dgram-multicast-set-interface.js:
```diff
@@ -73,6 +73,20 @@ const dgram = require('dgram');
+      socket.close();
+    }));
+  } catch (e) {
+    common.skip(`IPv6 sockets are missing or incompatible: ${e}`);
```
There is `common.hasIPv6`, which lets you detect if IPv6 is supported before trying to do something like this.

lostnet commented:
Thanks to @yorkie and @cjihrig, I went with the common.hasIPv6 for the RPi issue.

- I'd like to see new Windows tests to know if my changes to bind helped there.
- For Linux, redhat based distributions with broken default filtering, against fedora I could reduce the tests but the redhat one seems to make it possible to sufficiently reduce the loopback test to pass.
- For Solaris based systems, I've gotten verification that no 5.11 based system is likely to have bug free multicast loopback. It will affect every 5.11 through an as of yet unreleased update.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/3733/

lostnet commented:
Thanks @jasnell for the CI tests, I'm not sure if the **arm-fanned** test, I think it failed to start due to /tmp being full?

- **windows** looks like it will live.

- I notice that the **fedora** tests are only failing on **rackspace nodes** and not on digitalocean ones, so I think either the setup scripts or their patch levels are different in a way that addresses iptables filtering local traffic. If they could be normalized in the direction of digital ocean's config that should address that problem.

- For the **linux-one**,  I notice that it is its own architecture so I could exclude an os.arch() of s390[x]? It would be nice if someone could verify that this setup has an explanation like a firewall. (AFA my changes, the test shows that its failure is with multicast in general and not setting the interface.)

jasnell commented:
@mhdawson ... Would you be able to help figure out what's happening with the Linux one failure here?

lostnet commented:
## linux one
Looking at the build repository, I've found an explanation for linuxone:
[iptables: enable loopback](https://github.com/nodejs/build/blob/b3f563d73a1d777362a4578931301ff702338c3f/setup/rhel72-linuxonecc/ansible-playbook.yaml#L43)

This would imply to me that nothing works by default. So I could create a new rule to add there.


## fedora/rackspace

But on digitalocean vs rackspace, I see no iptables manipulation in the fedora setups so I think slightly different nodes are being provisioned from the two providers. I.e. server vs desktop?

lostnet commented:
@jasnell , @jbergstroem  & @mhdawson I opened [build pull request 477](https://github.com/nodejs/build/issues/477) with the changes needed to be able to do loopback multicast CI tests. Since it is a bit of an investment to test multicast support and likely to interact with other projects like [fedora 24 support](https://github.com/nodejs/build/pull/446), I'm not sure if it makes sense to hold this change up until it makes sense to work on that support or just move the comprehensive tests to one of the non-CI test directories?
thanks,
Will

mhdawson commented:
Sorry just going to issues I've been mentioned on and found this one I'd noticed before.  I'll try to sync up with @jbergstroem to discuss how we want to handle it.  I know we did have to adjust the firewall when enabling the tests initially so its a good questions as to whether we should have done that through ansible or in the management panel like I did in that case.

jasnell commented:
What do we want to do with this one?

jasnell commented:
Ping @lostnet ... is this still something you'd like to pursue?

ping @mhdawson @mcollina @jbergstroem ... any thoughts?

lostnet commented:
The remaining test failures all seem to be explained by default packet filter rules not exempting local multicast. I created a build pull request based on the changes fedora needed:

https://github.com/nodejs/build/pull/477

I'm not aware of anything else I can do without either help making progress on integrating multicast support in build (ideal) or manual intervention by someone with build access to change the packet filter rules and rerun my tests for this pull request (and I would need to move the test to a non-CI directory.)

mcollina commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,84 @@ added: v0.6.9
+interface. All packets sent to multicast on the socket will be sent on the
+interface determined by the most recent successful use of this call.
+
+For IPv6 sockets, `multicastInterface` should include a scope to indicate the
```
I do not think that "should" is the right term. here and in the previous paragraph.

lostnet replied:
I've no idea in terms of informal English. In [RFC terminology](https://www.ietf.org/rfc/rfc2119.txt), Line 315 has the only actual MUST, leaving some strong recommendations that I would call SHOULD:

In the IPv4 case, I mean SHOULD in the sense that any OS MAY be choosing other behaviors for other IPs that one should not rely on for maximum portability.
In the IPv6 case, I mean SHOULD as in it is technically optional to add the default scope, but I would recommend including it specifically to avoid bugs through miscommunication.

mcollina commented on doc/api/dgram.md:
```diff
@@ -299,6 +299,84 @@ added: v0.6.9
+
+A socket's address family's ANY address (IPv4 `'0.0.0.0'` or IPv6 `'::'`) can be
+used to return control of the sockets default outgoing interface to the system
+for future multicast packets.
```
This whole section is not in harmony with the rest of the document. I would try to sum it up a bit.
Also, I would move the content _before_ the examples.

lostnet replied:
I added this level of detail in response to @jasnell 's request and then some requests for clarification of terms. Given that combinations like windows, IPv6 and multicast are extreme edge cases but need the most detail to implement correctly, I don't think it is possible to create a harmony where more common cases get the most (or at least equal) attention. I also don't see a clear way to restructure multicast related methods to a separate doc?

mcollina commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,288 @@
+    process.exit(1);
+  }, TIMEOUT);
+
+  // Launch the child processes.
```
why this test have to fire child processes?

sam-github replied:
I have same question, it seems completely unnecessary. If you do fork children, the children should exit if the parent does to prevent leaking: `process.on('disconnect', () => process.exit(1))`

lostnet replied:
If changes to the parent of this test:
> cjihrig commented on Aug 2, 2016
> Perhaps you could add a test similar to test/internet/test-dgram-multicast-multi-process.js.

are requested and made before this pull is complete then I will also integrate them in this request.

jasnell commented:
Ping. What do we want to do with this one?

mcollina commented:
I LGTMed it, so it can land with some nits, cleaning up the unit test, a fresh CI run etc.

@lostnet are you still willing to push this through? Sorry for making this process last so much.

lostnet commented:
Hi @mcollina I'd be happy to resolve the docs conflict and then try to rework my docs changes to fit in better.

The area where I really got stuck was the build configuration for local multicast testing on OSes with broken defaults. I suggested [some example changes](https://github.com/nodejs/build/pull/477) but I need to know if changes will be made and if not, if/how I can reduce the tests and still integrate.

mcollina commented:
If the problem is only a build one, then we can merge this and skip the test on those platforms until the issue is resolved.

cc @nodejs/build

lostnet commented:
Hi @mcollina I've added updates to pass changes to automated checks since last year. I think there are other [changes](https://github.com/nodejs/node/issues/14016) that have happened since then that I should probably integrate.

mcollina commented:
@lostnet what do you plan to integrate?

CI: https://ci.nodejs.org/job/node-test-pull-request/9850/

lostnet commented:
@mcollina  I resolved the merge conflict in my fork in [b44b469] (https://github.com/nodejs/node/pull/7855/commits/b44b46905da776fb62942fe61205b095aea0bee8) but that was probably the wrong way to do it given that the CI looks like it is applying the commits in order via a rebase. If I understand the collaborator guide correctly, I think I need to rebase to nodejs/master?

mcollina commented:
You can open a new PR if you want, or just force push to master with a rebased commit.

BridgeAR commented:
This got rebased. I think this could use some more eyes @nodejs/collaborators

mcollina commented on lib/dgram.js:
```diff
@@ -563,6 +563,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+
+  const err = this._handle.setMulticastInterface(interfaceAddress);
+  if (err) {
+    throw errnoException(err, 'setMulticastInterface');
```
It should use the new error system

mcollina commented on lib/dgram.js:
```diff
@@ -563,6 +563,19 @@ Socket.prototype.setMulticastLoopback = function(arg) {
+  this._healthCheck();
+
+  if (typeof interfaceAddress !== 'string') {
+    throw new TypeError('interfaceAddress must be a string');
```
it should use the new error system

jasnell commented:
@mcollina ... no, since those errors are within a brand new method this would be semver-minor. The semver-major requirement is only for converting existing errors over to the new system.

lostnet commented:
@mcollina & @jasnell I've changed the TypeError to the new system. For the `errnoException` it seems like _errnoException was modified to match api/errors.md in this [change](https://github.com/nodejs/node/commit/58831b2f24a6b71b172e5f5876cb77096408142e) so I'm not sure if I need to change anything?

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9944/

mcollina commented:
I think this is good to go. Let's have a CI run before landing.

mcollina commented:
This is faling on CentOS and Red Hat because of `test/parallel/test-dgram-multicast-set-interface-lo`, with output:

```
pid 26011 iface 23.253.249.127 MULTICAST 224.0.0.116
pid 26002 iface 0.0.0.0 MULTICAST 224.0.0.115
outgoing iface 23.253.249.127
pid 26019 iface 127.0.0.1 MULTICAST 224.0.0.117
pid 26036 iface 23.253.249.127 MULTICAST 224.0.0.116
pid 26039 iface 127.0.0.1 MULTICAST 224.0.0.117
pid 26025 iface 0.0.0.0 MULTICAST 224.0.0.115
1504606017977 - 1504606017977
[PARENT] sent '1504606017977 - First message to send' to 224.0.0.115:12346
1504606017977 - 1504606017977
[PARENT] sent '1504606017977 - Second message to send' to 224.0.0.115:12346
1504606017977 - 1504606017977
changing outgoing multicast 23.253.249.127
[PARENT] sent '1504606017977 - Third message to send' to 224.0.0.116:12346
1504606017977 - 1504606017977
[PARENT] sent '1504606017977 - Fourth message to send' to 224.0.0.117:12346
1504606017977 - 1504606017977
[PARENT] sent '1504606017977 - Fifth message to send' to 224.0.0.116:12346
1504606017977 - 1504606017977
changing outgoing multicast 127.0.0.1
[PARENT] sent '1504606017977 - Sixth message to send' to 224.0.0.117:12346
[PARENT] sendSocket closed
[CHILD] 26039 received '1504606017977 - Sixth message to send' from {"address":"127.0.0.1","family":"IPv4","port":12346,"size":37}
[CHILD] 26019 received '1504606017977 - Sixth message to send' from {"address":"127.0.0.1","family":"IPv4","port":12346,"size":37}
[PARENT] 26039 received 1 messages total.
[PARENT] 26019 received 1 messages total.
[PARENT] Responses were not received within 5000 ms.
[PARENT] Fail
```

Do you know what might be the problem?

lostnet commented:
@mcollina Yes, that looks like [this](https://github.com/nodejs/build/pull/477), anything that got its default iptables from redhat doesn't identify local traffic based on actual origin and just passes traffic with between loopback addresses. I could probably provide a cleaner solution than I suggested in 477 using fwmark, but I've no idea what was supported in the oldest redhats under LTS..

mcollina commented:
@lostnet can you soft-detect the problem and skip the test if that's the case?

lostnet commented:
@mcollina I can't reallly differentiate loss of packets due to a filter from a bug relating to multicast except partially by adding new unicast pretests. I could make the test skip on lost packets:

```
$ git diff
diff --git a/test/parallel/test-dgram-multicast-set-interface-lo.js b/test/parallel/test-dgram-multicast-set-interface-lo.js
index 039e9a00fa..740b175d69 100644
--- a/test/parallel/test-dgram-multicast-set-interface-lo.js
+++ b/test/parallel/test-dgram-multicast-set-interface-lo.js
@@ -88,9 +88,10 @@ if (process.argv[2] !== 'child') {
   timer = setTimeout(function() {
     console.error('[PARENT] Responses were not received within %d ms.',
                   TIMEOUT);
-    console.error('[PARENT] Fail');
+    console.error('[PARENT] Skip');
 
     killChildren(workers);
+    common.skip('Check filter policy');
 
     process.exit(1);
   }, TIMEOUT);

```

allowing this skip only for `common.isLinux` seems to be as fine grained as I can go to capture all redhat derivatives?

mcollina commented:
@lostnet let's skip on lost packets. It's the best course of action - udp can lose packets anyway.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10065/

lostnet commented on test/parallel/test-dgram-multicast-set-interface-lo.js:
```diff
@@ -0,0 +1,293 @@
+      PORTS[msg.mcast],
+      msg.mcast,
+      function(err) {
+        assert.ifError(err);
```
If I force an EPERM error here with an iptables OUTPUT filter, the children processes don't get cleaned up. But, I don't think it needs to be addressed now, since this would show up as the first failure in a run it affects and is not different than [the test this is based on](https://github.com/nodejs/node/blob/master/test/internet/test-dgram-multicast-multi-process.js).

lostnet commented:
Thank you @mcollina I think enough comparison data will build up in CI pass/skip results for this test to help investigate any problems reported on specific OSes later.

I've also added one code line comment for the only error I was able to cause in my testing, but I don't expect to see it in CI as I don't think any OS would ship with OUTPUT filters.

mcollina commented:
@cjihrig can you have a look? I'm 👍  and this is ready to land, CI is green.

mcollina commented:
Landed as 4ae0afb12b6d29c8bb743c4fd560a91fd626b558.

Thanks @lostnet for the contribution! It has taken some time to brew but it was worth it!
